### PR TITLE
Resolve and record builds in database

### PR DESCRIPTION
**Pull request by chuazhan**
Add new method to resolve odcs compose id for flatpak
image.

Add logic to filter out compose in sets renew and new.
Link composes to build and record builds in database.

JIRA: RHELWF-3780

### CODE REVIEW

**Comment by chuazhan** [comment]
@qixiang @hluk @gnaponie Please help to review, thanks!

File: tests/handlers/koji/test_rebuild_flatpak_application_on_module_ready.py [hunk 1]
```diff
@@ -178,5 +180,172 @@ def test_event_state_updated_when_no_images_with_higher_rpm_nvr(self):
...
         )
+
+    def test_get_module_name_stream_set(self):
+        build = {
+            "build_id": 1850907,
+            "epoch": None,
+            "extra": {
+                "typeinfo": {
+                    "module": {
+                        "modulemd_str": '---\ndocument: modulemd\nversion: 2\ndata:\n  name: ghc\n  stream: "9.2"\n  version: 3620211101111632\n  context: d099bf28\n  summary: Haskell GHC 9.2\n  description: >-\n    This module provides the Glasgow Haskell Compiler version 9.2.1\n',
```
Comment (gnaponie) [inline tests/handlers/koji/test_rebuild_flatpak_application_on_module_ready.py:193]: could you put this string on multiple lines for readability?

Comment (chuazhan) [reply tests/handlers/koji/test_rebuild_flatpak_application_on_module_ready.py]: Ok. Fixed.

File: freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py [hunk 2]
```diff
@@ -167,3 +179,152 @@ def _filter_images_with_higher_rpm_nvr(self, event, rebuild_images_list):
         )
         images = lb.get_images_by_nvrs(rebuild_images_list, rpm_nvrs=errata_rpm_nvrs)
         return images
+
+    def _get_module_name_stream_set(self, event):
```
Comment (qixiang) [inline freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py:183]: There is `self.event` can be used to get the advisory id, unnecessary to have `event` in api params.

File: freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py [hunk 3]
```diff
@@ -167,3 +179,152 @@ def _filter_images_with_higher_rpm_nvr(self, event, rebuild_images_list):
...
+                ArtifactType.IMAGE,
+                state=ArtifactBuildState.PLANNED.value,
+                original_nvr=nvr,
+                rebuild_reason=RebuildReason.DIRECTLY_AFFECTED.value,
+            )
+            # Set context to particular build so logging shows this build
+            # in case of error.
+            self.set_context(build)
+
+            module_name_stream_set = self._get_module_name_stream_set(event)
```
Comment (qixiang) [inline freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py:290]: For every image, you're calling this API and the underline errata API once which is unnecessary.

File: freshmaker/lightblue.py [hunk 4]
```diff
@@ -467,6 +467,34 @@ def resolve(self, lb_instance, children=None):
             err = "Cannot resolve the container image: %s" % e
             self.log_error(err)
 
+    def resolve_flatpak_odcs_compose_ids(self):
```
Comment (qixiang) [inline freshmaker/lightblue.py:470]: There is not much different between this API and `resolve_original_odcs_compose_ids`, could you just add a param to the original API:
```
def resolve_original_odcs_compose_ids(self, from_most_original_build=True):
    ...
    original_nvr = self.nvr
    if from_most_original_build:
        original_nvr =  ArtifactBuild.get_most_original_nvr(self.nvr)
    ...
```

File: tests/handlers/koji/test_rebuild_flatpak_application_on_module_ready.py [hunk 5]
```diff
@@ -178,5 +180,174 @@ def test_event_state_updated_when_no_images_with_higher_rpm_nvr(self):
...
+        get_build.return_value = build
+        get_cve_affected_build_nvrs = self._patch(
+            "freshmaker.errata.Errata.get_cve_affected_build_nvrs"
+        )
+        get_cve_affected_build_nvrs.return_value = ["ghc-9.2-3620211101111632.d099bf28"]
+        ret = self.handler._get_module_name_stream_set(self.event)
+        self.assertEqual(ret, {"ghc:9.2"})
+
+    @patch("freshmaker.odcsclient.create_odcs_client")
+    def test_prepare_data_for_compose(self, create_odcs_client):
```
Comment (gnaponie) [inline tests/handlers/koji/test_rebuild_flatpak_application_on_module_ready.py:219]: It seems like there's a lot of repeated code in this test and it becomes hard to read it. Any chance we can shorten this a bit? Consider using `parametrize` or maybe a for loop.

File: freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py [hunk 6]
```diff
@@ -167,3 +179,152 @@ def _filter_images_with_higher_rpm_nvr(self, event, rebuild_images_list):
...
+            for source in source_value.split():
+                group = source.strip().split(":")[:2]
+                name_stream_set.add(group[0] + ":" + group[1])
+
+            # If all the sources are not in advisory, renew them
+            if len(name_stream_set - module_name_stream_set) == len(name_stream_set):
+                renewed_odcs_compose_set.add(compose_id)
+            else:
+                # If some source in advisory, create composes for all separately
+                new_odcs_compose_set.update(name_stream_set)
```
Comment (qixiang) [inline freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py:244]: I believe odcs module compose source should be module nsvc instead of just name and stream, can you check this?

File: freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py [hunk 7]
```diff
@@ -33,14 +34,15 @@
...
+from freshmaker.odcsclient import create_odcs_client
 from freshmaker.pyxis import Pyxis
-from freshmaker.types import EventState
+from freshmaker.types import ArtifactType, ArtifactBuildState, EventState, RebuildReason
 
 
 class RebuildFlatpakApplicationOnModuleReady(ContainerBuildHandler):
     # Module ready means Flatpak module advisory is in QE status
-    # and all attach build is in signed status
+    # and all attach build are signed.
```
[the remark this answers is not in the document]
Comment (chuazhan) [reply freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py:45]: Ack. Fixing

File: freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py [hunk 8]
```diff
@@ -167,3 +179,152 @@ def _filter_images_with_higher_rpm_nvr(self, event, rebuild_images_list):
...
+        """
+        errata = Errata()
+        module_nvrs = errata.get_cve_affected_build_nvrs(event.advisory.errata_id, True)
+
+        module_name_stream_set = set()
+        with koji_service(
+            conf.koji_profile, log, login=False, dry_run=self.dry_run
+        ) as session:
+            for module_nvr in module_nvrs:
+                mmd = session.get_modulemd(module_nvr)
```
[the remark this answers is not in the document]
Comment (chuazhan) [reply freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py:200]: Actually I was confused with the [doc](https://fedora-modularity.github.io/libmodulemd/latest/modulemd-2.0-Modulemd.html#modulemd-read-packager-string) too. But when I was testing it, the returned value is a _ResultTuple like "(<GType ModulemdModuleStreamV2 (94271507489296)>, object=<Modulemd.ModuleStreamV2 object at 0x7fe3eea26700 (ModulemdModuleStreamV2 at 0x55bd4bb78940)>)"

File: freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py (hunk 2 above)
Comment (chuazhan) [reply freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py:183]: Ack. Fixing

File: freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py (hunk 3 above)
Comment (chuazhan) [reply freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py:290]: Sure. Fixing.

File: freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py (hunk 6 above)
Comment (chuazhan) [reply freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py:244]: Yes. When I use method GET, I can see the source format is nvsc, but I checked the [doc](https://pagure.io/odcs/#:~:text=Requesting%20new%20compose), when we create a new compose, for module source type, the source can be "White-space separated NAME:STREAM or NAME:STREAM:VERSION of modules to include in compose." I am wondering if the VERSION should be created automatically.

File: freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py [hunk 9]
```diff
@@ -167,3 +179,152 @@ def _filter_images_with_higher_rpm_nvr(self, event, rebuild_images_list):
...
+            is a set of required data for creating new composes
+        :rtype: tuple
+        """
+        renewed_odcs_compose_set = set()
+        new_odcs_compose_set = set()
+        for compose_id in original_odcs_compose_ids:
+            compose = self.odcs.get_compose(compose_id)
+            source_type = compose.get("source_type")
+            # source_type of module composes is 2
+            if source_type != 2:
```
[the remark this answers is not in the document]
Comment (chuazhan) [reply freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py:229]: Yeah, looks like a magic number, but this code imitates the pulp compose which source_type is 4 and that doesn't exist in ArtifactType. Do we need to create a new one, looks like ComposeSourceType, @qixiang @gnaponie any idea?

File: tests/handlers/koji/test_rebuild_flatpak_application_on_module_ready.py (hunk 5 above)
Comment (chuazhan) [reply tests/handlers/koji/test_rebuild_flatpak_application_on_module_ready.py:219]: The data has a small difference. Let me separate the test cases for readability.

File: freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (qixiang) [reply freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py:229]: ODCS compose type is different from freshmaker `ArtifactType`, you can use the types from ODCS:
```
>>> from odcs.common.types import PungiSourceType
>>> PungiSourceType.MODULE
2
```

File: freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py (hunk 6 above)
Comment (qixiang) [reply freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py:244]: If you request compose with only module `name` + `stream`, I guess ODCS should get the latest modules from that `name` and `stream`, which can be different from the modules in the trigger advisory, so I believe at least `version` is also required, but I'm not sure about `context`, I'd suggest checking with ODCS team if that is not documented.

File: freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (chuazhan) [reply freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py:229]: Ack. Thanks

File: freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py (hunk 6 above)
Comment (chuazhan) [reply freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py:244]: Ok. Checked with ODCS people and as you mentioned name + stream get the latest modules. Adding version as well

File: freshmaker/lightblue.py (hunk 4 above)
Comment (chuazhan) [reply freshmaker/lightblue.py:470]: Ok. Try to fix it.

File: freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py [hunk 10]
```diff
@@ -167,3 +179,152 @@ def _filter_images_with_higher_rpm_nvr(self, event, rebuild_images_list):
...
+        """
+        Get set of modules with format NAME:STREAM in an advisory.
+
+        :param event FlatpakModuleAdvisoryReadyEvent: The event this handler
+            is currently handling.
+        :return: a set with item NAME:STREAM
+        :rtype: set
+        """
+        errata = Errata()
+        module_nvrs = errata.get_cve_affected_build_nvrs(event.advisory.errata_id, True)
```
[the remark this answers is not in the document]
Comment (chuazhan) [reply freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py]: Fix the former part, not quite sure about why to save the mapping between nvr and nsvc

File: tests/handlers/koji/test_rebuild_flatpak_application_on_module_ready.py [hunk 11]
```diff
@@ -178,5 +188,159 @@ def test_event_state_updated_when_no_images_with_higher_rpm_nvr(self):
...
+        self.assertEqual(ret, {"ghc:9.2:3620211101111632"})
+
+    @patch("freshmaker.odcsclient.create_odcs_client")
+    def test_prepare_data_for_compose_not_module_type(self, create_odcs_client):
+        odcs = create_odcs_client.return_value
+        # Test for compose source_type not module type
+        odcs.get_compose.return_value = {
+            "arches": "x86_64",
+            "id": 985716,
+            "owner": "auto/liver.englab.brq.redhat.com",
```
Comment (gnaponie) [inline tests/handlers/koji/test_rebuild_flatpak_application_on_module_ready.py:206]: could you make a fake url instead of "redhat" (since this is a public repo)?

**Comment by gnaponie** [comment]
I've just left one comment, but the change looks good to me +1

File: freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py [hunk 12]
```diff
@@ -167,3 +180,149 @@ def _filter_images_with_higher_rpm_nvr(self, event, rebuild_images_list):
...
+            name_stream_set = set()
+            source_value = compose.get("source", "")
+            for source in source_value.split():
+                group = source.strip().split(":")[:3]
+                name_stream_set.add(":".join(group))
+
+            # If all the sources are not in advisory, renew them
+            if len(name_stream_set - module_name_stream_version_set) == len(
+                name_stream_set
+            ):
```
Comment (qixiang) [inline freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py:240]: Could you explain on this? I can't understand why this means all sources are not in advisory. And `version` should not be included while checking this.

File: freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py [hunk 13]
```diff
@@ -167,3 +180,149 @@ def _filter_images_with_higher_rpm_nvr(self, event, rebuild_images_list):
...
+                renewed_odcs_compose_set.add(compose_id)
+            else:
+                # If some source in advisory, create composes for all separately
+                new_odcs_compose_set.update(name_stream_set)
```
Comment (qixiang) [inline freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py:244]: So when some sources are in advisory, you'll create new composes for all these old modules, right? I believe this is wrong, for example, there is old modules:

- module-a:stable:123
- module-b:el8:345

And now you have module-b:el8:350 in advisory, you should request 2 new module composes (in case you want to create one compose for each module):
- compose: module-a:stable:123
- compose: module-b:el8:350

compose for `module-b:el8:345` should be excluded, in your code, there will be 3 composes with all these modules included.

**Review by qixiang — CHANGES_REQUESTED** [review]
It overall looks good to me, but still have something need to be fixed, especially the code of getting new odcs module composes.

File: freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (qixiang) [reply freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py]: I forgot this, could you try it in our freshmaker image? I guess it could be caused by some different version of packages, but I didn't dig into this, error from what I tried with `quay.io/factory2/freshmaker:prod`:
```
>  _, mmd = Modulemd.read_packager_string(modulemd_str, None, None)
                                                                                                                         
TypeError: cannot unpack non-iterable ModuleStreamV2 object
```

[the remark this answers is not in the document]
Comment (chuazhan) [reply freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py]: Ack. Fixed.
Add root cause for others, seems rpm package install is different with pip install, rpm will provide overrides which will help to unpack and strip GType, but pip install will not, so the return value is different.

File: freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py [hunk 14]
```diff
@@ -167,3 +180,149 @@ def _filter_images_with_higher_rpm_nvr(self, event, rebuild_images_list):
...
+            source_type = compose.get("source_type")
+            if source_type != PungiSourceType.MODULE:
+                renewed_odcs_compose_set.add(compose_id)
+                continue
+
+            name_stream_set = set()
+            source_value = compose.get("source", "")
+            for source in source_value.split():
+                group = source.strip().split(":")[:3]
+                name_stream_set.add(":".join(group))
```
[the remark this answers is not in the document]
Comment (chuazhan) [reply freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py]: Ack. Fixed. Actually these part has been refactored

File: freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py (hunk 12 above)
Comment (chuazhan) [reply freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py]: Ok. As discussed offline, has change this part with intersection and difference, please check the latest codes

File: freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py (hunk 13 above)
Comment (chuazhan) [reply freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py]: Ack. Fixed

**Comment by chuazhan** [comment]
@qixiang @chuazhan @hluk Updated, please help to review

**Comment by chuazhan** [comment]
@qixiang @gnaponie @hluk Can you help to take a look? Thanks

File: freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py [hunk 15]
```diff
@@ -167,3 +180,156 @@ def _filter_images_with_higher_rpm_nvr(self, event, rebuild_images_list):
...
+            if source_type != PungiSourceType.MODULE:
+                renewed_odcs_compose_set.add(compose_id)
+                continue
+
+            name_stream_set = set()
+            mapping = dict()
+            source_value = compose.get("source", "")
+            for source in source_value.split():
+                source = source.strip()
+                name_stream = source.rsplit(":", 2)[0]
```
Comment (qixiang) [inline freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py:237]: Is it possible that the module compose was generated from `<name>:<stream>`/`<name>:<stream>:<version>` in old build?

File: freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py [hunk 16]
```diff
@@ -167,3 +180,156 @@ def _filter_images_with_higher_rpm_nvr(self, event, rebuild_images_list):
...
+            inter_section = name_stream_set.intersection(module_name_stream_set)
+            if not inter_section:
+                renewed_odcs_compose_set.add(compose_id)
+            else:
+                # If some module in advisory, create composes for all differences,
+                # inter_section part will be added in following step
+                difference = name_stream_set.difference(module_name_stream_set)
+                for item in difference:
+                    new_odcs_compose_set.add(mapping[item])
+            # All modules in advisory should create compose
```
Comment (qixiang) [inline freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py:252]: Why all modules in advisory should be enabled for all impacted images? If there are multiple modules in advisory, each module can impact different images, for example, an advisory has `module-a` and `module-b`, `module-a` only impacts `a-flatpak-image`, `module-b` only impacts `b-flatpak-image`, in this case, you only need to request `module-a` compose for image `a-flatpak-image`.

File: freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py (hunk 15 above)
Comment (qixiang) [reply freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py:237]: Ignore this, tested and verified odcs will expand the module to `n:s:v:c` even it's requested with just `n:s` or `n:s:v`.

File: freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py (hunk 16 above)
Comment (chuazhan) [reply freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py]: Ack. Fixed

**Comment by chuazhan** [comment]
@qixiang @hluk @gnaponie  Please help to review, thanks!

**Review by qixiang — APPROVED** [review]
LGTM

File: freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py [hunk 17]
```diff
@@ -57,20 +60,38 @@ def handle(self, event):
...
+                else "No images are impacted by the advisory. "
+            )
+            msg = f"{msg} message_id: {event.msg_id}"
             db_event.transition(EventState.SKIPPED, msg)
             db.session.commit()
             self.log_info(msg)
             return []
 
-        images_nvr = ",".join([rebuild_image.nvr for rebuild_image in rebuild_images])
+        images_nvr = ",".join([rebuild_image.nvr() for rebuild_image in rebuild_images])
```
Comment (hluk) [inline freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py:84]: I might be wrong, but this change looks like a typo and seems there are no tests to catch this - i.e. missing test where `_filter_images_with_higher_rpm_nvr()` returns non-empty list.

File: freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py [hunk 18]
```diff
@@ -98,72 +119,219 @@ def _get_requests_session(cls, retry_options={}):
...
             server_url=conf.lightblue_server_url,
             cert=conf.lightblue_certificate,
             private_key=conf.lightblue_private_key,
             event_id=self.current_db_event_id,
         )
-        images = lb.get_images_by_nvrs(rebuild_images_list, rpm_nvrs=errata_rpm_nvrs)
+        images = lb.get_images_by_nvrs(images, rpm_nvrs=errata_rpm_nvrs)
         return images
+
+    def _prepare_data_for_compose(
```
Comment (hluk) [inline freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py:189]: This function is a bit difficult to understand. Can you split this into two function, each with simpler return type (`_outdated_composes()`, `_missing_composes()`)?

File: freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py [hunk 19]
```diff
@@ -98,72 +119,219 @@ def _get_requests_session(cls, retry_options={}):
...
+                            compose_source, "module"
+                        )
+                        db_compose = Compose(odcs_compose_id=compose["id"])
+                        db.session.add(db_compose)
+                        db.session.commit()
+                        odcs_cache[compose_source] = db_compose
+
+                    if db_compose:
+                        build.add_composes(db.session, [db_compose])
+                        db.session.commit()
```
Comment (hluk) [inline freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py:331]: Can there be just a single DB commit at the end of this method?

File: freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py (hunk 17 above)
Comment (chuazhan) [reply freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py]: Good catch. yeah, It is a typo. I updated the test_record_builds a bit to cover this.

File: freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py (hunk 18 above)
Comment (chuazhan) [reply freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py]: Ok. I split it in two methods,  but it seems it will lose some efficiency.

File: freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py (hunk 19 above)
Comment (chuazhan) [reply freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py:345]: I am assuming after every db operation, it will do a commit. But I am not sure we will strictly following this.

File: freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py [hunk 20]
```diff
@@ -98,72 +120,232 @@ def _get_requests_session(cls, retry_options={}):
...
+                    name_stream_set.add(name_stream)
+                    mapping[name_stream] = name_stream_version
+
+                inter_section = name_stream_set.intersection(module_name_stream_set)
+                if inter_section:
+                    # If some module in advisory, create composes for all differences,
+                    # inter_section part will be added in following step
+                    difference = name_stream_set.difference(module_name_stream_set)
+                    for item in difference:
+                        missing_composes.add(mapping[item])
```
Comment (hluk) [inline freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py:245]:
```suggestion
                missing_composes.update(
                    mapping[name_stream]
                    for name_stream in name_stream_set.difference(module_name_stream_set)
                )
```

File: freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py [hunk 21]
```diff
@@ -98,72 +120,232 @@ def _get_requests_session(cls, retry_options={}):
...
+        for compose_id in original_odcs_compose_ids:
+            compose = self.odcs.get_compose(compose_id)
+            source_type = compose.get("source_type")
+            if source_type == PungiSourceType.MODULE:
+                name_stream_set = set()
+                mapping = dict()
+                source_value = compose.get("source", "")
+                for source in source_value.split():
+                    source = source.strip()
+                    name_stream = source.rsplit(":", 2)[0]
```
Comment (hluk) [inline freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py:234]: Can you move this duplicate code to a separate function?

File: freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py [hunk 22]
```diff
@@ -98,72 +120,232 @@ def _get_requests_session(cls, retry_options={}):
...
-            for module_nvr in module_nvrs:
-                mmd = session.get_modulemd(module_nvr)
+            for advisory_module_nvr in self.advisory_module_nvrs:
+                mmd = session.get_modulemd(advisory_module_nvr)
                 content_index_url = "{}/released/contents/modules/{}:{}.json".format(
                     conf.flatpak_server_url,
                     mmd.get_module_name(),
                     mmd.get_stream_name(),
                 )
                 req_session = self._get_requests_session()
```
Comment (hluk) [inline freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py:149]: This should initialized once outside of the loop.

File: freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py [hunk 23]
```diff
@@ -98,72 +120,232 @@ def _get_requests_session(cls, retry_options={}):
...
+            name_stream_set = set()
+            source_value = compose.get("source", "")
+            for source in source_value.split():
+                source = source.strip()
+                name_stream = source.rsplit(":", 2)[0]
+                name_stream_set.add(name_stream)
+
+            # If all the sources are not in advisory, renew them
+            inter_section = name_stream_set.intersection(module_name_stream_set)
+            if not inter_section:
```
[the remark this answers is not in the document]
Comment (chuazhan) [reply freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py:213]: Ok. We don't need to create a new set for checking.

File: freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py (hunk 21 above)
Comment (chuazhan) [reply freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py:234]: Ack. Fixed

File: freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py (hunk 20 above)
Comment (chuazhan) [reply freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py:245]: I think we still need to check if the two sets are joint. I updated the codes, you can have a look with the new codes and see if it is proper.

File: freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py (hunk 22 above)
Comment (chuazhan) [reply freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py:149]: advisory_module_nvr is different, so every advisory_module_nvr will be initialized, right?

Comment (hluk) [reply freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py:149]: What I meant is that creating `requests.Session()` every loop iteration should be avoided.

File: freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py [hunk 24]
```diff
@@ -98,72 +120,256 @@ def _get_requests_session(cls, retry_options={}):
...
             server_url=conf.lightblue_server_url,
             cert=conf.lightblue_certificate,
             private_key=conf.lightblue_private_key,
             event_id=self.current_db_event_id,
         )
-        images = lb.get_images_by_nvrs(rebuild_images_list, rpm_nvrs=errata_rpm_nvrs)
+        images = lb.get_images_by_nvrs(images, rpm_nvrs=errata_rpm_nvrs)
         return images
+
+    def _parse_compose_source(self, source_value, name_stream_only=True):
```
Comment (hluk) [inline freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py:195]: This function hides too many things again. It can be also free function since it does not depend on `self`.

Too simplify you can return/yield the split sources from compose object. So at call sites you would have:

```python
name_stream_set = {f'{n}:{s}' for n,s,v in _compose_sources(compose)}
...
mapping = {f'{n}:{s}':f'{n}:{s}:{v}' for n,s,v in _compose_sources(compose)}
name_stream_set = set(mapping.values())
```

The main things to de-duplicate and move to separate function/method are:
- Getting the sources from compose object
- Splitting into n:s:v

**Review by hluk — CHANGES_REQUESTED** [review]

File: freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py [hunk 25]
```diff
@@ -98,72 +120,256 @@ def _get_requests_session(cls, retry_options={}):
...
+        for compose_id in original_odcs_compose_ids:
+            compose = self.odcs.get_compose(compose_id)
+            source_type = compose.get("source_type")
+            if source_type != PungiSourceType.MODULE:
+                outdated_composes.add(compose_id)
+                continue
+
+            name_stream_set, _ = self._parse_compose_source(compose.get("source", ""))
+
+            # If all the sources are not in advisory, renew them
```
Comment (hluk) [inline freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py:236]: No need for comment - the code is clear.

```suggestion
```

File: freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py [hunk 26]
```diff
@@ -98,72 +120,256 @@ def _get_requests_session(cls, retry_options={}):
...
+        """
+        missing_composes = set()
+        for compose_id in original_odcs_compose_ids:
+            compose = self.odcs.get_compose(compose_id)
+            source_type = compose.get("source_type")
+            if source_type == PungiSourceType.MODULE:
+                name_stream_set, mapping = self._parse_compose_source(compose.get("source", ""), False)
+
+                # If the two sets are joint, union compose's name_stream_version set
+                # and advisory module's name_stream_version set in missing composes.
```
Comment (hluk) [inline freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py:265]: No need for comment - it explains what the code does instead of why - the code is clear enough.

```suggestion
```

File: freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py (hunk 22 above)
Comment (chuazhan) [reply freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py:149]: Oh. You mean line "req_session = self._get_requests_session()"? I was thinking the difference parts.

File: freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py (hunk 25 above)
Comment (chuazhan) [reply freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py:236]: Ok. I was thinking to emphasize advisory, but module_name_stream_set's meaning can be seen in doc string.

File: freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py (hunk 26 above)
Comment (chuazhan) [reply freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py:265]: Ack. Removed

File: freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py (hunk 22 above)
Comment (hluk) [reply freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py:149]: Yeah, it something we missed in previous PR.

File: freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py (hunk 24 above)
Comment (chuazhan) [reply freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py]: Got it. Updated

**Review by hluk — APPROVED** [review]
LGTM! :+1:
